express-notes-api/index.js

```diff
@@ -0,0 +1,85 @@
+const express = require('express');
+const app = express();
+const fs = require('fs');
+const notesJSON = require('./data.json');
+
+app.use('/', express.json());
+
+function rewriteNotes(notes) {
+  const updatedNotes = JSON.stringify(notes, null, 2);
+  fs.writeFile('data.json', updatedNotes, 'utf-8', function (err, res) {
+    if (err) {
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+      process.exit(1);
+    }
+  });
+}
+
+app.get('/api/notes', (req, res) => {
+  const notesArr = [];
+  for (const id in notesJSON.notes) {
+    notesArr.push(notesJSON.notes[id]);
+  }
+  res.status(200).json(notesArr);
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const noteKey = parseInt(req.params.id);
+  if (Number.isInteger(noteKey) === false || noteKey < 0) {
+    res.status(400).json({ error: 'ID is not a positive integer' });
+  } else if (notesJSON.notes[noteKey] !== undefined) {
+    res.status(200).json(notesJSON.notes[noteKey]);
+  } else {
+    res.status(404).json({ error: 'Note cannot be found at inputted id' });
+  }
+});
+
+app.post('/api/notes', (req, res) => {
+  const newNote = req.body;
+
+  if (newNote.content === undefined) {
+    res.status(400).json({ error: 'Note must have a content property' });
+  } else {
+    newNote.id = notesJSON.nextId;
+    notesJSON.notes[notesJSON.nextId] = newNote;
+    notesJSON.nextId++;
+    rewriteNotes(notesJSON);
+    res.status(201).json(newNote);
```

Comment:
This doesn't wait to successfully write the data before reporting success. As I mentioned above, i recommend just moving your `fs.writeFile` call into here so you can include an `if/else` in its callback.